pg_search/src/postgres/storage/utils.rs

```diff
@@ -265,45 +282,64 @@ impl Drop for BM25BufferCache {
 
 type NeedsLock = bool;
 pub struct BufferMutVec {
-    inner: Vec<(pg_sys::Buffer, NeedsLock)>,
+    inner: [(pg_sys::Buffer, NeedsLock); MAX_BUFFERS_TO_EXTEND_BY],
+    cursor: usize,
 }
 
 impl BufferMutVec {
-    pub fn new(buffers: Vec<(pg_sys::Buffer, NeedsLock)>) -> Self {
-        Self { inner: buffers }
+    pub fn new(buffers: [(pg_sys::Buffer, NeedsLock); MAX_BUFFERS_TO_EXTEND_BY]) -> Self {
+        Self {
+            inner: buffers,
+            cursor: Default::default(),
+        }
     }
 
     /// Claim a buffer from the start, which ensures that the buffers are in the same order as they were created.
     /// Typically this means in order of increasing block number.
-    pub fn claim_buffer(&mut self) -> Option<pg_sys::Buffer> {
-        if self.inner.is_empty() {
-            None
-        } else {
-            let (buffer, needs_lock) = self.inner.remove(0);
-            if needs_lock {
-                unsafe {
-                    pg_sys::LockBuffer(buffer, pg_sys::BUFFER_LOCK_EXCLUSIVE as i32);
-                }
+    pub fn next(&mut self) -> Option<pg_sys::Buffer> {
+        if self.cursor >= MAX_BUFFERS_TO_EXTEND_BY {
+            return None;
+        }
+
+        let (buffer, needs_lock) = self.inner[self.cursor];
+        self.cursor += 1;
+
+        if buffer == pg_sys::InvalidBuffer as pg_sys::Buffer {
+            return None;
+        }
+
+        if needs_lock {
+            unsafe {
+                pg_sys::LockBuffer(buffer, pg_sys::BUFFER_LOCK_EXCLUSIVE as i32);
             }
-            Some(buffer)
         }
+        Some(buffer)
     }
 }
 
 impl Drop for BufferMutVec {
     fn drop(&mut self) {
-        for (buffer, needs_lock) in self.inner.drain(..) {
+        loop {
+            if self.cursor >= MAX_BUFFERS_TO_EXTEND_BY {
+                break;
+            }
+
             unsafe {
-                if buffer != pg_sys::InvalidBuffer as pg_sys::Buffer
-                    && pg_sys::InterruptHoldoffCount > 0
-                    && crate::postgres::utils::IsTransactionState()
+                let (buffer, needs_lock) = self.inner[self.cursor];
+                if buffer == pg_sys::InvalidBuffer as pg_sys::Buffer {
+                    break;
+                }
+
+                if pg_sys::InterruptHoldoffCount > 0 && crate::postgres::utils::IsTransactionState()
```

Comment:
Can these checks be moved out of the loop entirely? I know that the latter can't change under us... and if the former can change during the loop, then there is a race condition anyway.